### PR TITLE
Cache access to ProjectHDFio

### PR DESCRIPTION
**Pull request by pmrv**
Sketch of cached HDF5 access.  In case of loading a bunch of structures from an atomistic job with `get_structure()` as [here](https://github.com/pyiron/pyiron_atomistics/pull/189), this affords a 10x speed up.

### CODE REVIEW

**Comment by pmrv** [comment]
I replaced the standard library cache with an LFU cache from [cachetools](https://github.com/tkem/cachetools) with a default size limit of 8GiB.

File: .ci_support/environment.yml [hunk 1]
```diff
@@ -4,6 +4,7 @@ dependencies:
   - coveralls
   - coverage
   - codacy-coverage
+  - cachetools =4.2.2
```
Comment (liamhuber) [inline .ci_support/environment.yml:7]: Seeing the 100% coverage badge on this library
please

**Review by liamhuber — COMMENTED** [review]
Very cool. A few questions:

(1) Is the 8GB shared across a single universal cache, or is it 8GB/cached function? The latter seems excessive.

(2) What is the behaviour when the cache is smaller than a single instance of the returned value?

(3) Can ~~we~~EDIT: you(😉) get the demo implemented as a simple unit test? Also testing (2) and verifying (at least implicitly through timing stuff) that the cache clearing is working as expected?

**Comment by pmrv** [comment]
> Very cool. A few questions:
> 
> (1) Is the 8GB shared across a single universal cache, or is it 8GB/cached function? The latter seems excessive.

Per function, I can change it.  I just want to make sure that even large structures lists etc. get cached.  But yeah, it's on the larger side.

> (2) What is the behaviour when the cache is smaller than a single instance of the returned value?

It just won't end up in the cache.

> (3) Can ~we~EDIT: you(wink) get the demo implemented as a simple unit test? Also testing (2) and verifying (at least implicitly through timing stuff) that the cache clearing is working as expected?

Sure, will get that in.

**Comment by jan-janssen** [comment]
I am not exactly sure if this works for our setup. If I have a job running in non modal mode and I have a second process accessing the the job, how does the cache recognize the file changed? I guess that can be tested with a simple function which reads from a file, a separate process is writing to, to see if the cache is recognizing this.

**Comment by pmrv** [comment]
If a second job updates/adds an HDF entry in the file that does not reside in the cache, we're good.

If a second job updates an HDF entry in the file, while it still resided in the cache, that would indeed result in incoherence.  I guess this would be a problem for interactive jobs, if another job is processing them while they're still running.

I could invalidate the cache if the 'last-modified` file stamp changes between function calls, but that's maybe a bit round about.  Alternatively we optimize the method to be faster.  Something definitely has to change, because a factor 10 slowdown because our HDF layer is weird is not acceptable.

**Pull request closed without merging**